Port `addcmul` kernels to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62319 Port `addcdiv` to structured kernels.
* **#62318 Port `addcmul` kernels to structured kernels.**

Tracking issue: #55070

This PR introduces the method `TensorIteratorBase::build_ternary_op` for building a
`TensorIteratorBase` for 3-input 1-output kernel.

Differential Revision: [D29961997](https://our.internmc.facebook.com/intern/diff/D29961997)

ysiraichi committed:
Port `addcmul` kernels to structured kernels.

Tracking issue: #55070

This PR introduces the method `TensorIteratorBase::build_ternary_op` for building a
`TensorIteratorBase` for 3-input 1-output kernel.

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/PointwiseOps.cpp:
```diff
@@ -9,39 +9,26 @@
-    const Tensor& tensor2,
-    const Scalar& value,
-    Tensor& result) {
-  checkBackend("addcmul_cpu", result, self.options().backend());
```
We're losing this, but it looks like it's redundant since TensorIterator defaults to checking that all input/output tensors are on the same device (which is equivalent to the backend check). We also want to deprecate `backend` at some point in the future anyway :)

ysiraichi replied:
Doesn't `set_output` checks that already?

bdhirsh replied:
yep!

bdhirsh commented:
xla failures aren't real, but the for_each test failure is real. You'll probably need to add/remove a `skipMeta()` to the corresponding tests in `test_foreach.py`

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29961997).

ysiraichi committed:
Update on "Port `addcmul` kernels to structured kernels."


Tracking issue: #55070

This PR introduces the method `TensorIteratorBase::build_ternary_op` for building a
`TensorIteratorBase` for 3-input 1-output kernel.

Differential Revision: [D29961997](https://our.internmc.facebook.com/intern/diff/D29961997)

[ghstack-poisoned]

ysiraichi commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -831,6 +831,15 @@ void TensorIteratorBase::build_comparison_op(const Tensor& out, const Tensor& a,
 
+void TensorIteratorBase::build_ternary_op(const Tensor& out, const Tensor& a,
+    const Tensor& b, const Tensor& c) {
+  build(TensorIteratorConfig()
```
I agree with you. Those config make sense, at least for how `addcmul` and `addcdiv` should work. Not completely sure whether it would safe to do so, though.

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29961997).
